### ISSUE TITLE
VeraCrypt 1.23->1.24u6 UEFI Boot Order Breakage?

### ISSUE THREAD

**Issue by metal450** [veracrypt/VeraCrypt#612]
Previously, I was using VeraCrypt 1.23 (on Windows 10, full system encryption).  My UEFI is configured to boot Grub, which I then use to chainload the VeraCrypt bootloader.

Today, I updated to 1.24-Update6.  After the update, I reconfigured the UEFI boot list to leave only the Grub entry, as usual.  However, since the update, it's no longer behaving as expected.

* If I reboot & hit F12 to see the boot list, everything looks correct: there's only one entry: Grub. If I select it, it shows the Grub menu, I can select the VeraCrypt entry, and boot Windows.
* However, if I'm in Windows and I reboot, it goes straight to the Veracrypt bootloader - not to Grub.  I have no idea why it's started doing this, as there's one single entry in my uefi boot list: Grub. There's no entry to directly boot to the VeraCrypt bootloader...yet that's what it's doing.
* If I ctrl+alt+delete to reboot a second time, now it goes to Grub - as it's supposed to.
* Likewise, if I reboot from Linux, my 2nd OS, it goes straight to Grub - as it's supposed to.

In other words, the first boot/reboot after leaving Windows always ignores the UEFI boot list, and just goes straight to VeraCrypt. Other boots/reboots work properly/as configured.

I should mention that I've updated Veracrypt many times since I've had this Grub->VeraCrypt configuration, and never had any issue, so I can only assume that something was changed in the upgrade that broke it...?

**Comment by metal450** [comment]
Addendum: since it looks like nobody is going to address this, I tried to just rollback VeraCrypt to 1.23 to get it working properly again.  But it won't allow me to downgrade. So I'm now stuck not being able to boot Grub without doing multiple reboots every time (or doing a full system decrypt/re-encrypt, which is many many terabytes)...

Is there no way to get a response about this? It looks like the vast majority of issues posted on here are just ignored...

**Comment by dhammel** [comment]
I cant really answer your original question as I have little to no experience with veracrypt system encryption, as I use linux not windows.

I am pretty sure in order to downgrade veracrypt version, system must not be encrypted. Meaning you would need to decrypt to do this, same as if you wamted to uninstall it.

As for what I do know about veracrypt system encryption, it does have same casual issues with bootloaders on certain devices, depending on the manufactuer and how its delt with in BIOS. Do you have another computer or 2 to try it on? Are you dual booting?

**Comment by metal450** [comment]
>>I am pretty sure in order to downgrade veracrypt version, system must not be encrypted. Meaning you would need to decrypt to do this, same as if you wanted to uninstall it.

Right, exactly what I meant by "doing a full system decrypt/re-encrypt, which is many many terabytes." It would be quite a shame to have to decrypt & re-encrypt many terabytes of data, which would take ages & put needless wear on the SSDs, simply because it looks like a bug might have been introduced & the developers of the software aren't addressing issue reports :/

>>it does have some casual issues with bootloaders on certain devices, depending on the manufactuer and how its delt with in BIOS.

But per above, it was working just fine on 1.23 for well over a year - only upon upgrading to 1.24 did it break. So I don't see how it could be caused by the device/BIOS when I know for a fact that it had been working on this device/BIOS.

**Comment by dhammel** [comment]
Its possible its a bug, no software, or developer is perfect. Its also possible you may have done something that may not have known it to cause an issue. Are you sure you mean terabytes and not gigabytes? SSD's with many many terabytes would be costly and I havent seen any with that much, just usually around 3tb. However I dont use SSD's because of its limited write cycle, so I may be off or just havent looked throughly.

Have you done a BIOS update at all? Could try that as well, if havent. Could be an issue with a windows update, we all know how crummy windows updates can be, one of my top like 100 reasons I dont use windows. Boot order management can be done within windows, amd even linux to perfer a certain bootloader like GRUB to appear each time. I once set that on linux so it appears everytime, even in a non-dualboot environment.

It looks to me like windows is the culprit, as issues only occurs after booting windows, and rebooting windows. You can check the system configuration by press WIN+R. In the run box type: msconfig.exe
When it opens go to the Boot tab and see if you can adjist settings acording to your needs and see if windows will stop defaulting to its os. I dont have windows to test this on, so you may just have to trial and error. Before changing settings, see if the settings/information shown looks to be the reason why its booting straight into windows.


In order to downgrade though, you might have to decrypt fully unless theres some project maybe here on github that someone came up with to avoid the need but I dont know how that would even be done, and may even cause issues if done incorrectly.

**Comment by metal450** [comment]
>>Its possible its a bug, no software, or developer is perfect.

Absolutely - I too am a developer (of both FOSS & commercial software). No judgement or issue whatsoever with unintended bugs cropping up.  The only issue is not being able to get replies.

>>Its also possible you may have done something that may not have known it to cause an issue.

All I did was run the VeraCrypt installer, reboot, & set the UEFI boot order.  Couldn't be simpler. I've done it countless times for previous VeraCrypt updates.

>>Are you sure you mean terabytes and not gigabytes?

Yes, I'm sure. 4tb SSDs are widely available.

>>I dont use SSD's because of its limited write cycle

Exactly why (among other reasons) completely decrypting and re-encrypting seems like an unnecessary & not-so-great idea, when a simple discussion with the developers could possibly lead to a proper solution.

>>Have you done a BIOS update at all?

Nope.  I changed one thing - VeraCrypt version - and it broke. I wanted to avoid changing any other variables & thus complicating the issue further.

>>Could be an issue with a windows update, we all know how crummy windows updates can be

I have a firewall that blocks windows updates on the connection level, so definitely not it.

>>It looks to me like windows is the culprit, as issues only occurs after booting windows, and rebooting windows. 

...But Windows hasn't changed. Only VeraCrypt changed. This could also be phrased as "issues only occur after rebooting from the Veracrypt-encrypted OS, and only when the Veracrypt bootloader is >1.23." I don't see how Windows could be the culprit when the same copy of windows, unchanged, had been working until the exact precise boot cycle in which Veracrypt was updated. Only one variable changed when the behavior changed, and it wasn't Windows.

**Comment by dhammel** [comment]
When you mentioned many many tbs I was thinking 80-100+ (why i thought you meant GBs), 4tb is more understanding. 

I can understand not wanting to put extra stress on an ssd.

From what you mention, i see how looks to be a bug in veracrypt. Maybe that after the update it set the windows os as default (Could be an issue that just needs a one time manual adjustment) Did you try msconfig?
If msconfig doesnt work, there are most likely tools our their that can change boot order management on a more deeper level. I remember playing with one a few years ago when performing dual boot, before dumping windows all together.

If none of this works, a downgrade may be the only soultion to determine for sure if its a bug in veracrypt, and if so to await a fix.
I havent seen any reports of this issue from others.

You can even go to the more social forum for this software discussion on sourceforge.
If you want, go to veracrypt.fr/en and click Fourms on top right, there you can post an issue, in case im of no help, you have a higher chance of getting replies for this software.

**Comment by metal450** [comment]
>>Maybe that after the update it set the windows os as default (Could be an issue that just needs a one time manual adjustment)

Yup, it does change the UEFI boot order, which I did the one-time adjustment as mentioned in the OP: "After the update, I reconfigured the UEFI boot list to leave only the Grub entry, as usual."

>>Did you try msconfig?

I'm pretty sure that would only be relevant once the uefi has already started down the path of the windows bootloader, which would only happen once veracrypt has already decrypted windows, which only happens once it's decided to ignore Grub being first in the UEFI list & instead go straight to veracrypt.  msconfig is much later in the process than where this issue is, as I understand it.

>>there are most likely tools our their that can change boot order management on a more deeper level.

You might be thinking of BOOTICE, a common way to change uefi boot order from within windows (or from a PE). It effectively does the same thing as going to the BIOS & editing the order from there.

>>You can even go to the more social forum for this software discussion on sourceforge.

I've tried posting issues on sourceforge multiple times in the past - unfortunately, just like here, they were always ignored by the developers :(

**Comment by dhammel** [comment]
I believe VeraCrypt only has one primary developer working on VeraCrypt and one other mainly foe the bootloader. Mounir Idrassi in the past has had personal tkme which caused a big delay in the development.

The windows boot manager is located on a system reserved (hidden by default) partition, therefore can be accessed and loaded before os boots, veracrypt does not encrypt this partition, so it shouldnt matter. Boot management even be done on my luks encrypted installation, has no issues with modifying this from the linux cli. Looking at msconfig wouldnt hurt anything, messing with settings you dont know what they do could. If you got a backup of data, which you always should anyway could try anything at this point to attempt to fix it, at least I would but im just that dangerous lol

Can you look at msconfig boot section and tell me if its got both your os's displayed? If it does, does it label windows as default?

Windows, just like linux should be able to detect any other os instance regardless of encryption. Grub for example is unencrypted and would display your linux os details in it to know how to boot linux, windows can see this.

Ive messed around a lot with grub and default os's, just not from windows that much.

**Comment by metal450** [comment]
Grub is installed on the (unencrypted) EFI system partition.  It's the only "main" bootloader, the only one that knows about the other OSs (it has a custom menu I wrote from scratch that chainloads either VeraCrypt (aka encrypted Windows), or Linux.  Note: this is not the "standard" linux grub menu, it's my own 1st level bootloader that chainloads either of the others.

msconfig doesn't know about any of the other bootloaders or OSs.  No Windows boot menu is shown because Windows has no need/reason to know about the previous bootloader from which it was chainloaded.

If the EFI were first invoking the grub bootloader - as it had before the update - it would work. Since the update, rather than respecting the EFI boot order & booting straight to grub, it's as if it's "remembering" the chainloaded veracrypt bootloader, ignoring UEFI, & just booting straight back to that.  Windows doesn't know about grub in the same way that Linux doesn't know it's encrypted by VeraCrypt: it's at a "later" level.  Usually you have VeraCrypt->Windows. I have Grub->VeraCrypt->Windows. Nothing in that list knows about the item before it. If I do a cold boot, it still works.  But since the update, it skips the 1st Grup part and just goes straight to the VeraCrypt part.

**Comment by dhammel** [comment]
If your not using the official grub, it would be harder for me to diagnose it then. I dont have a similar setup at all to you, there is some issue somewhere within the windows side, and way boot management is handled. With my computer, i just change the boot order within linux and all is done, even with similar issues to you, just only on a single boot system. Might have to try posting on bigger forums, to get attention. Ive had issues like this, and some took me hours to fix, others days..If you had a traditional dual boot system, msconfig should have done it to resolve this.

Will probably have to start from the beginning to see what was actually changed during the veracrypt update, what the software alterted that is.

Did you try looking at the grub configuration manually to see if anyrhing is causing this?

On my side I can configure official grub or grub2, to default to linux, windows, or manual selection by ease. Works with an veracrypt encrypted windows, and luks encrypted linux install. I dont know what you wrote, how you wrote it and what for, so its hard for me to know what you may need to check for. Basically check to see if grub has any alterted configuration that may need updating.

**Comment by metal450** [comment]
Basically this is as simple as:

As of VeraCrypt 1.24u6, the first bootloader that the system boots to after a from-veracrypt reboot is not the first entry on the uefi boot list.

I should revise my previous statement - the order is UEFI->Grub->VeraCrypt->Windows.  UEFI has a list that selects which bootloader. Grub is 1st on that list, but since VeraCrypt 1.24, the 1st item on that list is no longer booted 1st. Veracrypt seems to prioritize rebooting to its own bootloader, bypassing the order in UEFI.

Everything else - msconfig, luks, etc - occurs later & is thus not relevant at that step of the boot process.

**Comment by dhammel** [comment]
ok, well ask yourself 'what would veracrypt have to update in order to prioritize its bootloader'

As its been a few years since using veraceypt system encryption, as its windows only I havent had to try this.

Have you looked at grubs configuration which you can access at /etc/default/grub on linux? I wish I knew more to help you, but this is just the way I manage grub and the way I would solve such an issue if I were on a dual boot setup, but as you have different modifications made, Im not sure if this would help.

In grubs configuration, you can force it to always show grubs menu, change items on the menu, change default timeout...if that would help at all, if its a bug in veracrypt thats persistent, then veracrypt may just change it back afterwards. In that case making it read-only would probably help, but also dangerous when comes to certain task.

**Comment by metal450** [comment]
>>ok, well ask yourself 'what would veracrypt have to update in order to prioritize its bootloader'

This is the key to this issue, but I'm really not sure. That's at an earlier stage of the boot process than I've dealt with.

All the rest of your answer again deals with grub. It never calls the grub bootloader at all.  What's in Linux's grub menu is irrelevant at this stage of the boot process, as Grub is never invoked - whether the grub menu would be hidden, would include veracrypt, etc has no effect if the grub bootloader never takes control of the boot process.

(And again, it's not Linux's default grub, it's the grub I manually installed on the EFI partition).

**Comment by dhammel** [comment]
When i referred to grub being hidden, basically stating that it may have just expired its timeout, booting default os, not displaying grub. Did you perform any test to know grub is not being invoked at all? or just assume being it goes straight to veracrypt bootloader?

Can you temporarily switch to the official grub, to make proper adjustments as a way to test it?

Is there some specific reason why you are using a special grub, that would make impossible to do?

**Comment by metal450** [comment]
I'm certain that grub is not timing out.

The Veracrypt entry isn't its default, there's no timeout set, & since its menu isn't even in the default location, there's no way Veracrypt would know to open its .cfg file, make edits, change it, and change it back (nor would veracrypt have a reason to do so).

This is why I'm using my own grub setup: https://superuser.com/questions/1090155/truecrypt-veracrypt-on-one-hdd-with-1-encrypted-1-unencrypted-win7-partition

**Comment by dhammel** [comment]
What method did you use to determine grub is not timing out?

If veracrypt is not able to update grub, then I couldnt see how a bug in veracrypt could cause this, as grub manages the setup and can be manually defined to which os to boot. Assuming that grub stays as priority 1 in UEFI boot

What steps have you actually tried to resolve this? Maybe itll give me a better idea, as to what you were figuring was wrong

**Comment by metal450** [comment]
>>I couldnt see how a bug in veracrypt could cause this, as grub manages the setup and can be manually defined to which os to boot. Assuming that grub stays as priority 1 in UEFI boot

Veracrypt seems to be doing something like saying "on next reboot, ignore the UEFI boot order and just boot straight to the veracrypt bootloader." I don't know how this works exactly, but it's definitely possible for software to do - i.e. think of cloning software that says "reboot your computer," and on next boot, it goes to their bootloader so it can do stuff like clone the system drive (without you having to manually add an entry to UEFI).  Like a one-time temporary change to the boot order.  Even BOOTICE has an option like this (https://lnk2.tk/snaps/2020-05-11_19.23.30.png).  Since the update, VeraCrypt seems to be doing something similar.

>>What method did you use to determine grub is not timing out?

Again, there's no timeout in the grub menu, there's no default option, and Veracrypt would have no way to know how to edit the .cfg file & then change it back (that wouldn't make any sense anyway - why in the world would veracrypt go to all the effort of searching for the location of a grub .cfg, spouting text into it, then searching it for that text and later removing it). It's just not something they're doing. There's zero reason for it, it isn't mentioned in the changelogs, it would take a lot of deliberate work on their part, it would be extremely fragile, & I'm not even using a "typical" grub setup.  The modification time of the grub menu file hasn't changed. They just aren't doing this.

>>What steps have you actually tried

1) Visit the firmware. Confirm there's only one entry in the UEFI boot list: grub.
2) Startup the computer. It boots to the grub menu, as expected.
3) Ctrl+Alt+Delete from the grub menu to reboot. Next startup again goes to grub, as expected.
4) Select the grub option to chainload a non-veracrypt OS, then reboot from that OS. Next startup again goes to grub, as expected.
5) Select the grub option to chainload the veracrypt bootloader, decrypt, boot the OS, then reboot. **Next startup goes straight to the Veracrypt bootloader** (and not grub, the only entry in the UEFI list)
6) Decrypt & boot the veracrypt OS again, then reboot. It goes straight to the Veracrypt bootloader (2nd time, just to reconfirm)
7) Decrypt & boot the veracrypt OS again, then reboot. This time, hit F12. In this pre-boot UEFI menu, I can see that grub is still the only entry. If I select grub this UEFI menu - aka explicitly telling it to use grub, per the uefi's only entry rather than ignoring the uefi menu & instead booting back to the veracrypt bootloader - the grub menu works.

This last fact says definitively that it's not grub that's changing. Explicitly telling it to boot to grub works. It's not a different or modified menu.  But unless you explicitly tell it to boot to grub, it ignores the order specified in the UEFI and goes straight to veracrypt.

Veracrypt is somehow overriding the uefi boot order for the 1st reboot out of the veracrypt-encrypted OS.

**Comment by metal450** [comment]
In the changelog for 1.24-Update2, I found two entries that are very suspicious (seem to directly relate to this behavior):

    * Fix multi-OS boot compatibility by only setting VeraCrypt as first bootloader of the system if the current first bootloader is Windows one.
    * Add new registry flags for SystemFavoritesService to control updating of EFI BIOS boot menu on shutdown.

I guess I'll try re-posting this on sourceforge in hopes that the developers actually respond to issues there.

**Comment by metal450** [comment]
Reported again here: https://sourceforge.net/p/veracrypt/tickets/346/

**Comment by dhammel** [comment]
>     * Fix multi-OS boot compatibility by only setting VeraCrypt as first bootloader of the system if the current first bootloader is Windows one.
>     * Add new registry flags for SystemFavoritesService to control updating of EFI BIOS boot menu on shutdown.
> 

It looks possible, good catch. Have you looked at th commits made jn the source to see if you can further track it down? Ill take a look soon, im also going to boot into a windows 10 vm soon as i can wjere k have verwcrypt installed for windows and check on something relating to it.

I havent seen a thread that the developer responsed to in a few months, so it seems he may be away for now. There are definately replies made to others, so hopefully someone woth the knowledge eill reply.

Good luck :)

**Comment by metal450** [comment]
Here are the commits:

https://github.com/veracrypt/VeraCrypt/commit/9741c9209d0deedcc5d60434e5e7df3c531ff192

https://github.com/veracrypt/VeraCrypt/commit/14a477026d6c9f3a549ba0dcc07955a8c70becfb

This makes me virtually *certain* it's something he broke in 1.24u2.

**Comment by dhammel** [comment]
I would referwnce that in any post you make online, looks like it is indeed possible.

If possible, I would say you could try building from source maybe since the latest commit prior to 1.24U2, but ive never overwritten an official binary with custom one built from source on windows so the issue of decrypting may still need to be done to 'downgrade'. May be more trouble than you want to go through now, but i figured in the time being...

**Comment by metal450** [comment]
Here's the problem & solution: https://superuser.com/questions/1506032/prevent-veracrypt-or-any-other-program-from-changing-the-order-of-boot-loaders

(I finally found it after manually digging through the source code, locating the registry key, & googling for that specifically.  How frustrating to spend >4 hours tracking down such an impactful change of behavior - which was barely documented, given no UI option, & which could've been solved in 1 second with proper documentation and/or if they didn't simply ignore issue reports on github.)

**Comment by dhammel** [comment]
Glad you got the fix, thats what i was just looking at within Windows 10 vm, was based on the changelog when it stated about registry flag. Since I dont have a dual boot system I was gonna have you try it, but glad you got it fixed. New updates of veraceypt may add it back, so keep that in mind. I havent read too much of the source code, in regards to this issue, so I dont know for sure if this will be changed or has been for the next stable release.

**Comment by alt3r-3go** [comment]
@metal450, now that you've found the solution, would you be willing to submit a documentation update PR for that? if not, feel free to reopen this issue, I'll try to find some time to do that myself (or to fix, depending on what the end-state fix idea was, haven't looked into that).

**Comment by metal450** [comment]
Reopened.

**Comment by idrassi** [comment]
The mechanism that makes VeraCrypt update the bootloader and boot entries can be disabled using the registry as indicated in the previous post: under `HKEY_LOCAL_MACHINE\System\CurrentControlSet\Services\VeraCryptSystemFavorites\`, the REG_DWORD `VeraCryptSystemFavoritesConfig` must be set to 1.

Once this is done, VeraCrypt will never try to update bootloder or modify boot entries. If you still have issues with boot entries after setting this registry key, then something else is messing with the boot entries.

You can check the code that manage this which is in the function `SystemFavoritesServiceCtrlHandler` in [src/Mount/Mount.c](https://github.com/veracrypt/VeraCrypt/blob/master/src/Mount/Mount.c#L9595). Below is an extract of the code :

```
	case SERVICE_CONTROL_STOP:
		SystemFavoritesServiceSetStatus (SERVICE_STOP_PENDING);

 		if (!(BootEncObj->ReadServiceConfigurationFlags () & VC_SYSTEM_FAVORITES_SERVICE_CONFIG_DONT_UPDATE_LOADER))
		{
			try
			{
				BootEncryption::UpdateSetupConfigFile (true);
				if (!BootEncStatus.HiddenSystem)
				{
					// re-install our bootloader again in case the update process has removed it.
					bool bForceSetNextBoot = false;
					bool bSetBootentry = true;
					bool bForceFirstBootEntry = true;
					uint32 flags = BootEncObj->ReadServiceConfigurationFlags ();
					if (flags & VC_SYSTEM_FAVORITES_SERVICE_CONFIG_FORCE_SET_BOOTNEXT)
						bForceSetNextBoot = true;
					if (flags & VC_SYSTEM_FAVORITES_SERVICE_CONFIG_DONT_SET_BOOTENTRY)
						bSetBootentry = false;
					if (flags & VC_SYSTEM_FAVORITES_SERVICE_CONFIG_DONT_FORCE_FIRST_BOOTENTRY)
						bForceFirstBootEntry = false;
					BootEncryption bootEnc (NULL, true, bSetBootentry, bForceFirstBootEntry, bForceSetNextBoot);
					bootEnc.InstallBootLoader (true);
				}
			}
			catch (...)
			{
			}
		}
```

I have added checkbox in the UI for System setting in order to control the various aspects of this mechanism. Belew is a screenshot. I have also uploaded a preview of the upcoming 1.24-Update7 that include this to [Sourceforge Nightly Builds folder.](https://sourceforge.net/projects/veracrypt/files/VeraCrypt%20Nightly%20Builds/)

![VeraCrypt_EFI_SysENC_Settings](https://user-images.githubusercontent.com/4698669/89189477-4eb21a00-d5a0-11ea-9efe-453f39ad2b18.png)
